### DISCUSSION TITLE
Choose a new API that requires a token for authorization

### DISCUSSION THREAD

**Discussion by toby-thanathip** [nimblehq/android-templates#434] [RFCs]
Related to: https://github.com/nimblehq/android-templates/pull/382

As of now, our project is using an API that doesn't require a token for authorization. 

Because we are planning to [add Authenticator to our templates](https://github.com/nimblehq/android-templates/issues/228), we need a token for this.

Now the main question is, which API should we use? Here are some options:

- ❌ Nimble Survey API (Bad option, because we don't want to expose answers to our code challenge)
- ❌ GitHub API (Bad option, because the token can only be generated from the developer settings, there's no API for it)
- ❌ OpenWeather API (Bad option, because they use a _fixed_ API key in the request)

---

Then it finally hit me that we require an API, which supports _OAuth_ 🙈 

1️⃣ There are [several options available](https://github.com/public-apis/public-apis), it's nice we can work with _a real API_, but there are some downsides:
- No flexibility or control, as we must follow their API documentation
- It requires storing an API key/secret/username/password, in order to generate a token 

2️⃣ The second option is to create a [mock server with Postman](https://learning.postman.com/docs/designing-and-developing-your-api/mocking-data/setting-up-mock/), sadly _this is not a real API_, but it has its benefits: 🚀 
- More flexibility and control
- It doesn't requires storing any sensitive information

3️⃣ The final option is, don't add `Authenticator` to our templates and keep the current API.

→ But if all the clients require it, then we _must_ add it anyways.

4️⃣ Any other option? I'm all ears

**Comment by ryan-conway** [comment]
@toby-thanathip Do we need to include the API in the template though? It seems pretty likely that the generated project is going to update their `BASE_URL` value 🤔 
Since this is a template, it seems like the Authenticator should be API agnostic, so we could use any API for proof of work but we probably don't need the API in the template itself, just the Authenticator 😄 What do you think? 💡

**Reply by luongvo** [reply]
@ryan-conway @toby-thanathip IMO, we need a base `Authenticator` to form the standard logic for the implementation with some TODO comments for the refresh token API requesting action, etc. The main purpose of having an initial Authenticator is to avoid missing/wrong implementation.

The initial Authenticator needs to contain:
- max attempt/retry logic (normally set to 3).
- make use of `synchronized(this) {}` when executing the refresh token and re-build the request to handle multiple authenticator calls at a time.
- try/catch wrapping to catch errors during token refreshing. The default action in case an error has occurred.

**Reply by luongvo** [reply]
I added these requirements back to the ticket https://github.com/nimblehq/android-templates/issues/228 ✅

**Reply by luongvo** [reply]
Hence, I voted for `4️⃣ Other (explain)`

**Reply by luongvo** [reply]
The idea to have the initial Authenticator is the same as https://github.com/nimblehq/android-templates/discussions/522.
